File: built-in plugins dsp/mb-bass-upright.cpp
/**
 * MB Upright Bass
 * Category : instrument
 * Type     : bass
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Acoustic upright bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BASS_UPRIGHT_H
#define MB_BASS_UPRIGHT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBassUpright : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bass-upright";
    static constexpr const char* PLUGIN_NAME    = "MB Upright Bass";
    static constexpr const char* PLUGIN_TYPE    = "bass";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float body = 0.7f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbBassUpright() = default;
    ~MbBassUpright() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.body = std::clamp(params.body, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Upright Bass
        return input;
    }
};

#endif // MB_BASS_UPRIGHT_H

File: built-in plugins dsp/mb-bass-upright.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbBassUprightPlugin: PluginDefinition = {
    id: 'mb-bass-upright', slug: 'mb-bass-upright', name: 'MB Upright Bass', category: 'instrument', type: 'bass', version: '1.0.0',
    description: 'Acoustic upright bass', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'triangle', detune: 0, gain: 0.7 }, { type: 'sine', detune: 0, gain: 0.3 }],
    envelope: { attack: 0.02, decay: 0.5, sustain: 0.4, release: 0.5 },
    parameters: [
      { id: 'body', name: 'Body', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { body: 0.7, volume: 0.8 },
  };

export default MbBassUprightPlugin;
